### DISCUSSION TITLE
Compiled EXE crashes (exception) - it previously (apparently updated some modules) worked fine.

### DISCUSSION THREAD

**Discussion by KobiHaMitapes** [pyinstaller/pyinstaller#8639] [PyInstaller] [answered]
I compiled a python gui application using pyinstaller. Here is the compilation statement:
pyinstaller RoundTripDialog.py --add-data "icon/chromatic_small.ico;." --icon "icon/chromatic_small.ico" --splash splash.png -F --hidden-import lcms --hidden-import scipy --hidden-import scipy.interpolate --hidden-import scipy.spatial.cKDTree --hidden-import sklearn.metrics._pairwise_distances_reduction._datasets_pair --hidden-import sklearn.metrics._pairwise_distances_reduction._middle_term_computer --hidden-import scipy._lib._uarray._backend --name RoundTripTool --onefile

I get errors upon runing the executable - see the screenshot
![Exception](https://github.com/pyinstaller/pyinstaller/assets/93469471/51aefc94-42db-4898-a70f-6cb3d900f9aa)

Somehow _uarray is being found. In a previous compilation before I updated pyinstaller and modules, I was able to successfully run the executable. Has anyone experienced this before? Can anyone shed light on what i can to do correct this?
Thanks.

**Comment by rokm** [comment]
That looks like a missing or incompatible dependency of `scipy._lib._uarray._uarray` binary extension. Although the only dependencies I can see are the python shared library and api-ms-* DLLs, which should not be a problem on contemporary Win 10/11 systems.

Does it help if you rebuild with added `--clean` option?

What version of python, PyInstaller, and scipy are you using? Is this python.org + pip, or anaconda python environment?

Does the issue also occur if you freeze a simple
```python
import scipy._lib._uarray
print("hello")
```

program in onedir mode instead of onefile one?

If it does, can you archive build and dist directories (for this test program only) and upload the archive somewhere so I can take a look at it?

**Reply by KobiHaMitapes** [reply]
Thanks for your comment. I’m going to check it all at work. 
If I use the —clean, I just add it to the compilation command? 
regarding versions, I’ll answer again from that computer and let you know.  Here things without the array (apparently) compile and work. This did too.

**Reply by rokm** [reply]
>  If I use the —clean, I just add it to the compilation command?

Yes.

**Reply by KobiHaMitapes** [reply]
Now I'll try to answer your questions.
Python: 3.8.8 -- using the base environment - it was made for doing this project I'm doing.
I initially installed python using anaconda. Yes, because of things (I did not like particularly) I had to do install some modules using pip. We have stringent requirements at work and this is what resulted.  
pyinstaller               5.6.2            py38h2bbff1b_0         This appears to be conda
scikit-image              0.18.1                   pypi_0    pypi    Apparently pip                                                                    
scikit-learn                1.0.2                     pypi_0    pypi    Apparently pip                                                                 
scipy                         1.10.1                   pypi_0    pypi    Apparently pip

My latest compilation was:
 pyinstaller --clean RoundTripDialog.py --add-data "icon/chromatic_small.ico;." --icon "icon/chromatic_small.ico" --splash splash.png
-F --hidden-import lcms --hidden-import scipy --hidden-import scipy.interpolate --hidden-import scipy.spatial.cKDTree --hidden-import sklearn.metrics._pairwise_distances_reduction._datasets_pair --hidden-import sklearn.metrics._pairwise_distances_reduction._middle_term_computer --hidden-import scipy._lib._uarray._backend --hidden-import sklearn.utils._heap --hidden-import
 sklearn.utils._sorting --hidden-import sklearn.utils._vector_sentinel --hidden-import sklearn.neighbors._typedefs --hidden-import importlib_resources.trees --name HPSciTexRoundTripTool --onefile
 
 This (1) using the --clean like you suggested  (2) adding some more hidden imports based on some warnings/comments from compilations.
 
Sadly the uarray issue still arises - but ONLY in compilation. The python runs fine - the python always worked. The compilation worked as well until recently. 

If this helps you to understand what's going on, this will be great - if not, I can do other thinks like making a simple python project with just the uarray. I can understand that perhaps when compiling there is a mixup with uarray - these are typical compilation issues that throw exceptions. 

Please let me know what you think. Thank you very much.

**Reply by rokm** [reply]
>  pyinstaller 5.6.2 py38h2bbff1b_0 This appears to be conda

For starters, can you try not using out-of-date version of PyInstaller? If you have to use 5.x series, use 5.13.2...

Also, if you have stringent requirements, surely you know which modules you have updated, and you can try reverting those updates and see which one causes the issue? Particularly if you kept your PyInstaller pinned to an old version, which eliminates regressions on our part...

**Reply by KobiHaMitapes** [reply]
I think I did recently update the pyinstaller - it was part of something that I don’t remember. Sometimes I go so fast I forget all the details. I can most definitely upgrade my pyinstaller. 
I am not bound by any versions. I compiled several things as products in my company and they just need to compile and work. 
I can try 5.13. 
What version would you recommend based on the stuff you’ve seen? I can always revert back, unless with the change, conda or pip want to make a set of other changes as well. That’s what happened with my pyinstaller. 
if I need to install it via pip would that matter?

**Reply by bwoodsend** [reply]
> What version would you recommend based on the stuff you’ve seen?

Latest version of PyInstaller. No Conda at all (not even Python itself).

**Reply by KobiHaMitapes** [reply]
I noticed - sorry for not mentioning before, that there is something else in the list there called:
pyinstaller-hooks-contrib 2022.15            pyhd8ed1ab_0    conda-forge
Should this stay? Does it have any function?

**Reply by bwoodsend** [reply]
Yes

**Reply by KobiHaMitapes** [reply]
I can just keep the hooks contrib or should I uninstall it and reinstall a later version of it? 
sorry for being so step by step but I want to take the correct/recommended action.

**Reply by rokm** [reply]
If you can, you should start with clean python.org python installation, then create a clean virtual environment using `venv` module, and install your dependencies into it. Then install latest version of PyInstaller, which should also pull in the latest version of `pyinstaller-hooks-contrib`.

If you can't ditch conda, start with clean and minimal conda environment (not the base one). (So that if issues arise, you can give me the env file and I can recreate your environment locally).

**Reply by rokm** [reply]
Actually, one thing came to mind - since this is PyPI `scipy` and Anaconda python 3.8, it could be that the issue is due to delvewheel load order file not being collected from `scipy.libs` directory - we nominally added support for that in v5.6 (https://github.com/pyinstaller/pyinstaller/commit/fe4a77c52f9810a30e7dc62319a4764bd9f2ab2e), but I think the particular load order filename format used by scipy-1.10.1 PyPI wheel will be picked up only after v5.11.0 (https://github.com/pyinstaller/pyinstaller/commit/75f86ee17da619d4b138b1ee66eda78579d51122).

EDIT: on the second thought, that might not be the problem; the filename format seems to be using dash and not underscore, so it should be handled by 5.6 variant...

**Reply by KobiHaMitapes** [reply]
So I upgraded pyinstaller to 5.13.2 using pip. It was a bit interesting because I could not uninstall using conda so I have to take care of some issues with pathlib - but I overcame that. I ended up with the same issue. So, I made a simple project with, as you said:
```
import scipy._lib._uarray
print("hello")
```
Worked fine, in python.
I compiled it using:
`pyinstaller --clean uarrayTest.py --onedir --name uarrayTestCase`
Went into the directory created and ran the exe in command line and got:

![image](https://github.com/pyinstaller/pyinstaller/assets/93469471/5a51b431-9161-4992-a88e-de0c5d3ac129)

Please find a link, below,  to icloud to download the zip file. If all goes well, you'll find the executable and the rest of the package there. Perhaps you'll see my (probably very silly) mistake and can let me know what I need to do to fix this. The zip file is about 241mb - not sure what the download speed will be.  If this method doesn't work, let me know and I'll find another way to get the zip file to you.

[https://www.icloud.com/iclouddrive/047_lLL_3psUMZPEoPbH5Lp7Q#uarrayTestCase](url)

**Reply by bwoodsend** [reply]
> It was a bit interesting because I could not uninstall using conda so I have to take care of some issues with pathlib

Please use a clean environment. The sloppy broken residues of a butchered conda environment are exactly want we want to eliminate as a potential cause of all this.

**Reply by rokm** [reply, marked as answer]
Hmm, some of those libs are suspiciously small - are you perhaps using UPX during the build?

**Reply by KobiHaMitapes** [reply]
I’ll answer you both. The butchered conda seems not to be the issue only because it was the same before I upgraded the pyinstaller. 
yes there is upx and I am not sure I understand why. I don’t ask for it in my compile statement which I wrote above. Also in my real app the one file exe is larger than before when things worked with out the DLL not being found. 
how can I do it again and specify no upx?

**Reply by rokm** [reply]
Either remove UPX from PATH, or add [`--noupx`](https://pyinstaller.org/en/stable/usage.html#cmdoption-noupx) to your PyInstaller command.

**Reply by KobiHaMitapes** [reply]
Continuing the thread ... I removed the upx from the small test case and indeed it does now work.
Is it possible that the upx is the culprit? If so - yes you're right. I apparently so quickly put the upx into the path or as a variable (need to go through and see how i got it in there) and for the thing I was compiling then, it had no real effect. However, for this particular module, with the kd-tree, this behavior started.
I will try to compile the larger project and see whether it comes up now.  If it does, I'll mark it as the solution - albeit I am not sure this is what was intended with UPX ! But that is actually not that important to me. We can talk about the size and startup time at a different time. The most important thing is that if changes need to be made, that we can compile.

**Reply by bwoodsend** [reply]
I think we should consider making upx opt-in rather than opt out if UPX happens to be available.

**Reply by KobiHaMitapes** [reply]
I marked ROKM's comment as the answer. There were other things along the way - the pyinstaller 5.13 is different slightly than the previous version I used. I had to physically write 
import matplotlib
matplotlib.use('Qt5Agg')
where I didn't have to before. If effected everything - drag and drop that didn't behave well and plotting where it apparently thought I was using a different backend other than pyside2. 

I learned a lot. And I appreciate your kindness and patience - this was not an easy one. 
One question - you both seem to really prefer pip over using anaconda/conda. 
I'd like to know why this is the case. I always thought that conda was a better package manager, maintaining stability better. 
Currently in my company I'm finding SSL issues with conda relating, I guess, to networking security. Pip, on the other hand, works. So I need to adjust accordingly. But I'd like to know what professional reasons are there for preferring pip over conda. 
If you prefer to stay out of this - you can - I'm just trying to learn something and I am capable of changing, if necessary.

**Reply by bwoodsend** [reply]
> So I need to adjust accordingly. But I'd like to know what professional reasons are there for preferring pip over conda.

Conda do a lot of silly reshuffling of packages (e.g. trying to shoe-horn MKL into NumPy) and trying to make them vaguely look like relocatable UNIX style environments. Relocatable UNIX environments barely work on UNIX, let alone Windows, so they have to take a lot of messy shortcuts (e.g. using `PATH` to find libraries on Windows because there's no such thing as `@rpath`). Despite these very radical changes, Conda rarely bother testing their packages beyond a trivial `import library` so breakages are very common. There are even bugs in Conda's Python itself which affect PyInstaller -- most notably `os.add_dll_directory()` being a no-op, causing any package using delvewheel to have DLL load issues on Windows -- so even using Conda's Python with only pip packages is asking for trouble. They also have this triunality of conda packages + conda forge packages + pip packages. Not only can someone install a packages from mixed sources but Conda and pip will routinely trample packages from the other package manager, leaving environments in who knows what states. From a maintainer's point of view, this is hell because it multiplies the number of permutations we have to test and massively reduces the chance of us being able to reproduce an issue a user is reporting.

Specifically for PyInstaller (or any other packaging tool), the library shuffling and locating (e.g. the `PATH` thing on Windows) makes it very difficult to detect which libraries are needed by which packages but it also makes the environment very susceptible to masking those bugs caused by missing dependencies since, in the event that a library is not packaged properly, it can probably be pulled in from somewhere else on the machine it's being tested on. (There was a really nasty case of this with numpy segfaulting due to missing `libiomp5md.dll` which could not be reproduced on CI or most developer machines because various developer tools also provide this library. i.e. There's an inverse relationship between the probability of a person being able to see the bug and their being packaging-savvy enough to have any idea how to debug it.) Conda do have some package metadata which we can use to map packages to files but we can only loosely follow what it says because Conda ship all kinds of junk files (a lot of compiler debugger outputs) in their packages which we need to magically filter out or face the wrath of an indignant user who thinks it's PyInstaller's fault that their application is 50% rubbish.

All in all, I see zero benefit in using Anaconda. On a very very good day, your PyInstaller+Conda built package is 2-100x larger (yes 100 times larger!) than it needs to be but still work, on a realistic day, your application will be broken in a way that you won't be able to reproduce, the rest of the time, you'll get errors that don't help you and end up on our issue tracker. No developer tests with anaconda, nobody wants to test with anaconda because it would take a huge amount of work and CI time, and even if someone persevered and tested with anaconda anyway, they'd only catch a small subset of the conda inflicted problems.

**Reply by KobiHaMitapes** [reply]
Yes I do see that my packages are huge and in fact there was a package that I made as a separate exe because I thought maybe it would be worthwhile being an independent app as well. So in python, I just call subprocess and serialize the data into it etc. In the compiled version, instead of calling subprocessing another pyside2 python app, I subprocess the exe itself. Turns out that that silly exe is huge. As such it takes a long time to open and it's not really reasonable. 
That's how I got to UPX - I thought it might make the compiled subprocess open faster. But I was wrong! And I messed up other stuff - as you know - this is how all this began. 
If you're going to tell me this is because of conda and had I done things using pip only the executables would be smaller - I am willing to accept this and understand that from your very legitimate point of view, conda is a nuisance. I appreciate your frankness and all of your help. Indeed because of you guys, I was able to solve the problem (that I, in fact, caused myself).
I honestly can say I really learned a lot through this process.
